addons/mail/models/mail_channel.py

```diff
@@ -746,6 +747,36 @@ def _channel_message_notifications(self, message, message_format=False):
     #   - when a message is posted on a channel (to the channel, using _notify() method)
     # ------------------------------------------------------------
 
+    def toggle_message_pin(self, message_id, pinned):
+        """ (Un)pin a message on the channel and send a notification to the
+        members.
+        :param message_id: id of the message to be pinned.
+        :param pinned: whether the message should be pinned or unpinned.
+        """
+        self.ensure_one()
+        if not self.env["mail.message"].search([
```

Comment:
This is a bit confused ... Maybe we could benefits from the ORM and fields and do something like
```
message_toupdate = self.message_ids.filtered(
    lambda msg: msg.id == message_id and is_pinned == !pinned
)
if not message_toupdate:
    return
```